lightning-persister/src/lib.rs

```diff
@@ -0,0 +1,371 @@
+				assert_eq!(persisted_chan_data_0.keys().len(), 1);
+				for mon in persisted_chan_data_0.values() {
+					assert_eq!(mon.get_latest_update_id(), $expected_update_id);
+				}
+				persisted_chan_data_1 = persister_1.load_channel_data::<EnforcingChannelKeys>().unwrap();
+				assert_eq!(persisted_chan_data_1.keys().len(), 1);
+				for mon in persisted_chan_data_1.values() {
+					assert_eq!(mon.get_latest_update_id(), $expected_update_id);
+				}
+			}
+		}
+
+		// Create some initial channel and check that a channel was persisted.
+		let _ = create_announced_chan_between_nodes(&nodes, 0, 1, InitFeatures::known(), InitFeatures::known());
+		check_persisted_data!(0);
+
+		// Send a few payments and make sure the monitors are updated to the latest.
+		send_payment(&nodes[0], &vec!(&nodes[1])[..], 8000000, 8_000_000);
+		check_persisted_data!(5);
+		send_payment(&nodes[1], &vec!(&nodes[0])[..], 4000000, 4_000_000);
+		check_persisted_data!(10);
+
+		// Close the channel and make sure everything is persisted as expected.
+		// Force close because cooperative close doesn't result in any persisted
+		// updates.
+		nodes[0].node.force_close_channel(&nodes[0].node.list_channels()[0].channel_id);
+		check_closed_broadcast!(nodes[0], false);
+		check_added_monitors!(nodes[0], 1);
+
+		let node_txn = nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap();
+		assert_eq!(node_txn.len(), 1);
+
+		let header = BlockHeader { version: 0x20000000, prev_blockhash: Default::default(), merkle_root: Default::default(), time: 42, bits: 42, nonce: 42 };
+		connect_block(&nodes[1], &Block { header, txdata: vec![node_txn[0].clone(), node_txn[0].clone()]}, 1);
+		check_closed_broadcast!(nodes[1], false);
+		check_added_monitors!(nodes[1], 1);
+		check_persisted_data!(11);
+
+		fs::remove_dir_all("persister0").unwrap();
+		fs::remove_dir_all("persister1").unwrap();
+	}
+
+	// Windows ignores the read-only flag for folders.
+	#[cfg(not(target_os = "windows"))]
+	#[test]
+	fn test_readonly_dir() {
+		let persister = FilesystemPersister::new("persister".to_string());
+		let test_writeable = TestWriteable{};
+		let test_txo = OutPoint {
+			txid: Txid::from_hex("8984484a580b825b9972d7adb15050b3ab624ccd731946b3eeddb92f4e7ef6be").unwrap(),
+			index: 0
+		};
+		// Create the persister's directory and set it to read-only.
+		let path = &persister.path_to_channel_data;
+		fs::create_dir_all(path).unwrap();
+		let mut perms = fs::metadata(path).unwrap().permissions();
+		perms.set_readonly(true);
+		fs::set_permissions(path, perms).unwrap();
+		match persister.write_channel_data(test_txo, &test_writeable) {
+			Err(e) => {
+				assert_eq!(e.kind(), io::ErrorKind::PermissionDenied);
+			}
+			_ => panic!("Unexpected error message")
+		}
+		fs::remove_dir_all("persister").unwrap();
+	}
+
+	// Windows ignores the read-only flag for folders.
+	#[cfg(not(target_os = "windows"))]
+	#[test]
+	fn test_perm_failure_force_close() {
```

Comment:
Let me know if I have some misunderstanding here: Is there a need to use `FilesystemPersister` to test this behavior? Wouldn't it be more suitable to test this in the `lightning` crate's tests using a test persister? Similarly for `test_fail_on_open`.

Reply:
Well, the test persister doesn't try to write to disk, so this test wouldn't work with it as is. Similarly, with `test_fail_on_open`, the `TestPersister` doesn't attempt to create directories. Why would `lightning` be more suitable?

Reply:
My reasoning is that it is testing logic unrelated to `FilesystemPersister` (e.g., force closing a channel on persister error). It's just that it's using `FilesystemPersister` to trigger such logic whereas a test persister could have been used instead to achieve the same results.

That said, you've made think about this slightly differently than before. My new question would be why do we need to test this case at all? So long as we test that
1. a `Persister` error causes a `ChainMonitor` error (`lightning` unit (ideally) or functional test),
2. a `ChainMonitor` error causes a force close (`lightning` functional test), and
3. a write to disk failure causes an error in the `Persist` implementation for `FilesystemPersister` (`lightning-persister` unit tests), then

does it matter _how_ `ChainMonitor` returned a failure to test the force close? (1) and (2) are sufficient to cover this scenario whereas (3) ensures `FilesystemPersister` errors are correctly conveyed through the `Persist` trait.

Similarly, a user implementing their own persister should only need to do (3) -- write unit tests -- and be confident Rust Lightning will work as expected for the force close scenario.

Reply:
I'd argue the tests also serve as integration tests for the `FilesystemPersister`, but I see your point. And the tests sorta serve to document the cases where the `FilesystemPersister` isn't expected to work.

Sorry if dense but the action item I've taken away from this discussion is that it'd be good to add a strictly unit test for the `FilesystemPersister` (1 and 2 should already be implemented). Let me know what you think.

Reply:
> I'd argue the tests also serve as integration tests for the `FilesystemPersister`, but I see your point. And the tests sorta serve to document the cases where the `FilesystemPersister` isn't expected to work.
> 

My one pushback here would be that unit tests are more suitable for documenting `FilesystemPersister`'s expectations. They tend to have a simple _setup-action-assert_ structure that makes them short and easy to read. Additionally, when making code changes, a failing unit test helps pinpoint the bug whereas a dozen failing functional and integration tests makes it difficult to know what went wrong.

> Sorry if dense but the action item I've taken away from this discussion is that it'd be good to add (3). (1 and 2 should already be implemented). Let me know what you think.

Yeah, presumably just (3) would be needed. I didn't verify that (2) exists but assumed it may already. :)

Reply:
>My one pushback here would be that unit tests are more suitable for documenting FilesystemPersister's expectations.

Ah, that's a good point I hadn't considered. I switched all but one of the FilesystemPersister tests to be unit tests and left one integration test.

iiuc, for (2) `test_simple_monitor_permanent_update_fail` should have us covered.